applications/luci-app-libreswan/htdocs/luci-static/resources/view/libreswan/tunnels.js

```diff
@@ -0,0 +1,241 @@
+			addr[3] & (mask[3] >>> 0 & 255)
+		].join('.') + '/' +
+		network.maskToPrefix(mask.join('.'));
+}
+
+return view.extend({
+	load: function() {
+		return Promise.all([
+			network.getDevices(),
+			uci.load('libreswan'),
+		]);
+	},
+
+	render: function(data) {
+		var netDevs = data[0];
+		var m, s, o;
+		var proposals;
+
+		proposals = uci.sections('libreswan', 'crypto_proposal');
+		if (proposals == '') {
+			ui.addNotification(null, E('p', _('Proposals must be configured for Tunnels')));
+			return;
+		}
+
+		m = new form.Map('libreswan', 'IPSec Tunnels');
+
+		s = m.section(form.GridSection, 'tunnel');
+		s.anonymous = false;
+		s.addremove = true;
+		s.nodedescription = false;
+		s.addbtntitle = _('Add Tunnel');
+
+		o = s.tab('general', _('General'));
+		o = s.tab('authentication', _('Authentication'));
+		o = s.tab('interface', _('Interface'));
+		o = s.tab('advanced', _('Advanced'));
+
+		o = s.taboption('general', form.ListValue, 'auto', _('Mode'));
+		o.default = 'start';
+		o.value('add', _('Listen'));
+		o.value('start', _('Initiate'));
+
+		o = s.taboption('general', widgets.NetworkSelect, 'left_interface', _('Left Interface'));
+		o.datatype = 'string';
+		o.multiple = false;
+		o.optional = true;
+
+		o = s.taboption('general', form.Value, 'left', _('Left IP/Device'));
+		o.datatype = 'or(string, ipaddr)';
+		for (var i = 0; i < netDevs.length; i++) {
+			var addrs = netDevs[i].getIPAddrs();
+			for (var j = 0; j < addrs.length; j++) {
+				o.value(addrs[j].split('/')[0]);
+			}
+		}
+		for (var i = 0; i < netDevs.length; i++) {
+			o.value('%' + netDevs[i].device);
+		}
+		o.value('%defaultroute');
+		o.optional = false;
+		o.depends({ 'left_interface' : '' });
+
+		o = s.taboption('general', form.Value, 'leftid', _('Left ID'));
+		o.datatype = 'string';
+		o.value('%any');
+		o.modalonly = true;
+
+		o = s.taboption('general', form.Value, 'right', _('Remote IP'));
+		o.datatype = 'or(string, ipaddr)';
+		o.value('0.0.0.0');
+		o.value('%any');
+		o.optional = false;
+
+		o = s.taboption('general', form.Value, 'rightid', _('Right ID'));
+		o.datatype = 'string';
+		o.value('%any');
+		o.modalonly = true;
+
+		o = s.taboption('general', form.Value, 'leftsourceip', _('Local Source IP'));
+		o.datatype = 'ipaddr';
+		for (var i = 0; i < netDevs.length; i++) {
+			var addrs = netDevs[i].getIPAddrs();
+			for (var j = 0; j < addrs.length; j++) {
+				o.value(addrs[j].split('/')[0]);
+			}
+		}
+		o.optional = false;
+		o.modalonly = true;
+
+		o = s.taboption('general', form.Value, 'rightsourceip', _('Remote Source IP'));
+		o.datatype = 'ipaddr';
+		o.optional = false;
+		o.modalonly = true;
+
+		o = s.taboption('general', form.DynamicList, 'leftsubnets', _('Local Subnets'));
+		o.datatype = 'ipaddr';
+		for (var i = 0; i < netDevs.length; i++) {
+			var addrs = netDevs[i].getIPAddrs();
+			for (var j = 0; j < addrs.length; j++) {
+				var subnet = calculateNetwork(addrs[j].split('/')[0], addrs[j].split('/')[1]);
```

Comment:
Have you tested what happens here with an IPv6?

Reply:
This app version does not support IPv6, getIPAddrs only returns IPv4 addresses.

```javascript
        /**                                                                          
         * Query all IPv4 addresses of the logical interface.                        
         *                                                                           
         * @returns {string[]}                                                      
         * Returns an array of IPv4 addresses in CIDR notation which have been
         * registered by the protocol handler. The order of the resulting array
         * follows the order of the addresses in `ubus` runtime information.
         */
        getIPAddrs: function() {
```